[SPARK-18710][ML] Add offset in GLM

actuaryzhang opened the pull request:
## What changes were proposed in this pull request?
Add support for offset in GLM. This is useful for at least two reasons:

1. Account for exposure: e.g., when modeling the number of accidents, we may need to use miles driven as an offset to access factors on frequency.    
2. Test incremental effects of new variables: we can use predictions from the existing model as offset and run a much smaller model on only new variables. This avoids re-estimating the large model with all variables (old + new) and can be very important for efficient large-scaled analysis. 

## How was this patch tested?
New test.

@yanboliang @srowen @felixcheung @sethah

actuaryzhang committed:
add trait offset

actuaryzhang committed:
add offset setter

actuaryzhang committed:
implement offset in GLM

actuaryzhang committed:
add test for glm with offset

actuaryzhang committed:
minor cleanup

SparkQA commented:
**[Test build #71974 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71974/testReport)** for PR 16699 at commit [`a1f5695`](https://github.com/apache/spark/commit/a1f56952b4ffcdfb400ff0ce014987c50de5f33e).

SparkQA commented:
**[Test build #71974 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71974/testReport)** for PR 16699 at commit [`a1f5695`](https://github.com/apache/spark/commit/a1f56952b4ffcdfb400ff0ce014987c50de5f33e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71974/
Test FAILed.

actuaryzhang commented:
jenkins, retest this please

actuaryzhang committed:
add doc for GLRInstance

SparkQA commented:
**[Test build #71975 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71975/testReport)** for PR 16699 at commit [`d071b95`](https://github.com/apache/spark/commit/d071b95ccc94404d37cde7cc122cf8a13fd04449).

SparkQA commented:
**[Test build #71975 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71975/testReport)** for PR 16699 at commit [`d071b95`](https://github.com/apache/spark/commit/d071b95ccc94404d37cde7cc122cf8a13fd04449).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71975/
Test FAILed.

zhengruifeng commented:
you should not modify sharedParams directly. 
"// DO NOT MODIFY THIS FILE! It was generated by SharedParamsCodeGen."

And if there is no other algorithms inheriting hasoffset, I suggest that do not create this

imatiach-msft commented:
can you please:
1.) add [MLLIB] to the title
2.) fix the MiMa tests
The MiMa tests error seems to be related to the interface change, probably you need to add an exclude to the file "project/MimaExcludes.scala":

[error]  * abstract method getOffsetCol()java.lang.String in trait org.apache.spark.ml.param.shared.HasOffsetCol is inherited by class GeneralizedLinearRegressionBase in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("org.apache.spark.ml.param.shared.HasOffsetCol.getOffsetCol")
[error]  * abstract method offsetCol()org.apache.spark.ml.param.Param in trait org.apache.spark.ml.param.shared.HasOffsetCol is inherited by class GeneralizedLinearRegressionBase in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("org.apache.spark.ml.param.shared.HasOffsetCol.offsetCol")
[error]  * abstract synthetic method org$apache$spark$ml$param$shared$HasOffsetCol$_setter_$offsetCol_=(org.apache.spark.ml.param.Param)Unit in trait org.apache.spark.ml.param.shared.HasOffsetCol is inherited by class GeneralizedLinearRegressionBase in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("

actuaryzhang committed:
remove offset from shared param

SparkQA commented:
**[Test build #71994 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71994/testReport)** for PR 16699 at commit [`d2afcb0`](https://github.com/apache/spark/commit/d2afcb0c335c133f1a86c8587fd7459ac39935a0).

SparkQA commented:
**[Test build #71994 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71994/testReport)** for PR 16699 at commit [`d2afcb0`](https://github.com/apache/spark/commit/d2afcb0c335c133f1a86c8587fd7459ac39935a0).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71994/
Test FAILed.

actuaryzhang committed:
fix style issue

SparkQA commented:
**[Test build #71995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71995/testReport)** for PR 16699 at commit [`9eca1a6`](https://github.com/apache/spark/commit/9eca1a682d75e20df89ebdb0f1a01e02996f9c7f).

SparkQA commented:
**[Test build #71995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71995/testReport)** for PR 16699 at commit [`9eca1a6`](https://github.com/apache/spark/commit/9eca1a682d75e20df89ebdb0f1a01e02996f9c7f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71995/
Test PASSed.

actuaryzhang commented:
@zhengruifeng @imatiach-msft 
Thanks much for pointing out the issue due to the hasOffset trait. This is what caused the test to fail. I have moved it to the GLRBase class. Things are working now 👍

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/optim/IterativelyReweightedLeastSquares.scala:
```diff
@@ -20,6 +20,7 @@ package org.apache.spark.ml.optim
 import org.apache.spark.internal.Logging
 import org.apache.spark.ml.feature.Instance
 import org.apache.spark.ml.linalg._
+import org.apache.spark.ml.regression.GLRInstance
```
this is a bit strange - you are using the GLRInstance that you added to specifically GeneralizedLinearRegression.scala in the more generic optimizer IterativelyReweightedLeastSquares.scala.  It doesn't seem right for this file to depend on anything in the regression directory, it should really be the other way around.
I'm wondering if either:
1.) We can move the GLRInstance to a more generic place
2.) OR add it to Instance.scala as a separate case class eg "OffsetInstance"
3.) OR keep the offset a separate value

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -358,13 +384,13 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
-        val eta = model.predict(instance.features)
+    val reweightFunc: (GLRInstance, WeightedLeastSquaresModel) => (Double, Double) = {
+      (instance: GLRInstance, model: WeightedLeastSquaresModel) => {
+        val eta = model.predict(instance.features) + instance.offset
```
minor suggestion - instead of doing:
eta = prediction + offset
mu = fitted(eta)
newLabel = eta - offset + stuff
maybe do:
eta = prediction
mu = fitted(eta + offset)
newLabel = eta + stuff

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -736,15 +762,27 @@ class GeneralizedLinearRegressionModel private[ml] (
+      val eta = BLAS.dot(features, coefficients) + intercept
+      familyAndLink.fitted(eta)
+    } else {
+      throw new SparkException("Must supply offset value when offset is set.")
```
maybe a better error message might be:
Must supply offset to predict when offset column is specified

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -94,6 +94,17 @@ private[regression] trait GeneralizedLinearRegressionBase extends PredictorParam
+    "column name. If this is not set or empty, we treat all instance offsets as 0.0")
+
+  /** @group getParam */
+  def getOffsetCol: String = $(offsetCol)
```
it looks like you will need to update the validateAndTransformSchema method below to validate these parameters - eg check if the column exists? (similar to what the base class does for features/label columns)

imatiach-msft replied:
thanks for fixing!

imatiach-msft commented:
@actuaryzhang  added a couple comments, please take a look, thanks!

actuaryzhang committed:
rename to OffsetInstance and add param check

SparkQA commented:
**[Test build #72003 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72003/testReport)** for PR 16699 at commit [`d44974c`](https://github.com/apache/spark/commit/d44974cfe50092bb639a31aa7aa9b16eb1d21fae).

actuaryzhang commented:
@imatiach-msft Thanks so much for your detailed review. Incredibly helpful. I've addressed all your comments in the new commit. Major changes are highlighted below:
1. Create `OffsetInstance` in Instance.scala and remove dependency of IWLS on regression. 
2. Add param checking for offset. 

Let me know if there is anything else that needs improvement.

SparkQA commented:
**[Test build #72003 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72003/testReport)** for PR 16699 at commit [`d44974c`](https://github.com/apache/spark/commit/d44974cfe50092bb639a31aa7aa9b16eb1d21fae).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72003/
Test PASSed.

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -263,17 +288,21 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
     val model = if (familyObj == Gaussian && linkObj == Identity) {
       // TODO: Make standardizeFeatures and standardizeLabel configurable.
+      val wlsInstances: RDD[Instance] = instances.map { instance =>
+        Instance(instance.label - instance.offset, instance.weight, instance.features)
```
going over all OffsetInstance and converting again to Instance seems like it would be expensive and it would increase memory usage.  What do you think about the alternative of making OffsetInstance inherit from Instance - but then you would have to change Instance from a case class - and then just passing Instance to the fit below?  Is there anything else we could do here?

imatiach-msft replied:
another thing you can do, if you don't want to change the hierarchy, is to move the initialization of instances inside the if/else.  Then, for weightedleastsquares, you can just create RDD[Instance], but for IRLS weighted instances.

imatiach-msft replied:
the more I think about this code, it looks like moving the initialization of instances inside the if/else below and creating Instance in one case and OffsetInstance in the other would save some memory/time and wouldn't force you to mess with the Instance case class.

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/feature/Instance.scala:
```diff
@@ -27,3 +27,25 @@ import org.apache.spark.ml.linalg.Vector
+  }
+
+  /** Converts to an [[Instance]] object by leaving out the offset. */
+  private[ml] def toInstance: Instance = Instance(label, weight, features)
```
it looks like this method is only used once in a test case, might it be better to remove it?

imatiach-msft replied:
actually, another alternative might be to make OffsetInstance inherit from Instance (which I wrote below as well).  What do you think about this idea?

actuaryzhang replied:
Yes, this is only used once in the current code, and I can get rid of it. But I feel other regression-type models may use offset at some point and having this method will make it easier to switch between Instance and OffsetInstance.

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -736,15 +765,27 @@ class GeneralizedLinearRegressionModel private[ml] (
   private lazy val familyAndLink = new FamilyAndLink(familyObj, linkObj)
 
   override protected def predict(features: Vector): Double = {
-    val eta = predictLink(features)
```
it looks like if the family = Gaussian and link = Identity you don't even need to check for offsetCol here

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -736,15 +765,27 @@ class GeneralizedLinearRegressionModel private[ml] (
-  private def predictLink(features: Vector): Double = {
-    BLAS.dot(features, coefficients) + intercept
+  private def predictLink(features: Vector, offset: Double): Double = {
+    BLAS.dot(features, coefficients) + intercept + offset
```
it looks like in the case where family = Gaussian and link = Identity we shouldn't be adding offset?  But then, why would anyone set an offset column... maybe, to make sure this strange case does not happen, we can throw an exception in the validation logic if user accidentally set an offset column in the case that family = Gaussian and link = Identity?

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -753,14 +794,15 @@ class GeneralizedLinearRegressionModel private[ml] (
-    val predictLinkUDF = udf { (features: Vector) => predictLink(features) }
+    val predictUDF = udf { (features: Vector, offset: Double) => predict(features, offset) }
+    val predictLinkUDF = udf { (features: Vector, offset: Double) => predictLink(features, offset) }
+    val off = if (!isSet(offsetCol) || $(offsetCol).isEmpty) lit(0.0) else col($(offsetCol))
```
calling the offset column "off" was a little confusing to me, maybe we can give it a better name, like "offsetCol" or even "offset".  When I looked at the code below I first thought it was some sort of flag passed to the predict function.

imatiach-msft replied:
also, as I mentioned above, if family = Gaussian and link = Identity we shouldn't be passing any offset, but then the user shouldn't be setting an offset column probably - so either adding that to the validation logic or to the code here or to the predict function you call below would fix this unusual case; I think this seems like a validation issue and probably should be added to the validation method.

imatiach-msft commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -578,6 +578,79 @@ class GeneralizedLinearRegressionSuite
+      for (family <- Seq("gaussian", "poisson", "gamma")) {
+      val trainer = new GeneralizedLinearRegression().setFamily(family)
+        .setFitIntercept(fitIntercept).setOffsetCol("offset")
+        .setWeightCol("weight").setLinkPredictionCol("linkPrediction")
```
in my code reviews @jkbradley noted that unnecessary parameters should be removed.  It looks like most of these are already defaults and can be removed.

imatiach-msft commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -578,6 +578,79 @@ class GeneralizedLinearRegressionSuite
+      Vectors.dense(-1.9991044, 0.7247511, 0.1424392),
+      Vectors.dense(-0.27378146, 0.31599396, -0.06204946))
+
+    import GeneralizedLinearRegression._
```
should this be moved to the imports above?

imatiach-msft commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -578,6 +578,79 @@ class GeneralizedLinearRegressionSuite
+    var idx = 0
+    for (fitIntercept <- Seq(false, true)) {
+      for (family <- Seq("gaussian", "poisson", "gamma")) {
+      val trainer = new GeneralizedLinearRegression().setFamily(family)
```
it looks like the spacing is off a little here?  the val trainer = ... is at the same level as the for above.  I'm surprised the style checker didn't catch something like this.

imatiach-msft commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -578,6 +578,79 @@ class GeneralizedLinearRegressionSuite
+            val eta = BLAS.dot(features, model.coefficients) + model.intercept + offset
+            val prediction2 = familyLink.fitted(eta)
+            val linkPrediction2 = eta
+            assert(prediction1 ~= prediction2 relTol 1E-5, "Prediction mismatch: GLM with " +
```
I believe we are encouraged to use the ~== (triple signs)
At least, that is what I had to use in my PR for RandomForestSuite.scala

imatiach-msft commented:
This looks really good.  I made a couple more comments.  I'm not sure about the class hierarchy of Instance/OffsetInstance.  Maybe committers can take a look/comment on this PR? @jkbradley @srowen @yanboliang Thanks!

actuaryzhang committed:
create separate instance definition when initializing

actuaryzhang committed:
fix style in test

actuaryzhang commented:
@imatiach-msft Many thanks again for the review. I have incorporated some of your suggestions:
1. Create initialization of instance directly if it is Gaussian(identity) to avoid expensive conversion from OffsetInstance. 
2. Fix spacing issues in test
3. Change `off` to `offset` to be more informative. 

Feedback on the other suggestions:
1. `move import to beginning, use ~==`:  I did not change this for consistency with other tests in GLM. It's probably better to update all tests in a separate PR if this is preferred. 
2. Inheriting Instance: I'm not sure about this and leaning toward not touching `Instance` since it is used a lot elsewhere. Let's see what others suggest.  
Question: if `OffsetInstance` is a subclass of `Instance`, will a method with a signature `RDD[Instance]` also work with `RDD[OffsetInstance]`? 
3. Would you please elaborate on the following comment:
> if family = Gaussian and link = Identity we shouldn't be passing any offset.

This is the linear regression case and it still allows `offset`. The offset will be subtracted from the response as can be seen from the initialization in WLS `Instance(label - offset, weight, features)`.

SparkQA commented:
**[Test build #72040 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72040/testReport)** for PR 16699 at commit [`e183c08`](https://github.com/apache/spark/commit/e183c08373733249d7b9c476875c2a4a2dff3c05).

SparkQA commented:
**[Test build #72040 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72040/testReport)** for PR 16699 at commit [`e183c08`](https://github.com/apache/spark/commit/e183c08373733249d7b9c476875c2a4a2dff3c05).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72040/
Test PASSed.

actuaryzhang committed:
resolve conflict

SparkQA commented:
**[Test build #72067 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72067/testReport)** for PR 16699 at commit [`58f93af`](https://github.com/apache/spark/commit/58f93af6236d52f87c82411a645cf15413b30b9e).

actuaryzhang committed:
add test for tweedie

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
**[Test build #72068 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72068/testReport)** for PR 16699 at commit [`da4174a`](https://github.com/apache/spark/commit/da4174a6c2639001828e587794e1a32dcf8db15d).

SparkQA commented:
**[Test build #72068 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72068/testReport)** for PR 16699 at commit [`da4174a`](https://github.com/apache/spark/commit/da4174a6c2639001828e587794e1a32dcf8db15d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72068/
Test PASSed.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -336,14 +361,15 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
-        case Row(label: Double, weight: Double, features: Vector) =>
-          Instance(label, weight, features)
-      }
+    val off = if (!isDefined(offsetCol) || $(offsetCol).isEmpty) lit(0.0) else col($(offsetCol))
```
cast `offsetCol` to DoubleType here if you want to support non-double numeric datatype. 
`weightCol` and `labelCol` are automaticly cast into DoubleType in `Predictor.scala`

zhengruifeng replied:
BTW, the new params like `offsetCol` should be added into `Instrumentation.logParams` above.

imatiach-msft replied:
+1 on adding it to logParams, good catch

imatiach-msft commented:
@actuaryzhang thank you for updating the PR - I agree with your first two points.  With regards to the method of RDD[Instance] also working with RDD[OffsetInstance], I think it should work, as long as you define the method such that it takes in the parameter myAwesomeMethod(myRdd: RDD[_ <: Instance])
I think you need that "<:" in the method definition to allow the method to take in anything that inherits from Instance.

With regards to "Would you please elaborate on the following comment..." it looks like I made a mistake, I did not notice that you were subtracting the offset from the response when creating the Instance object in the previous commit, in which case you can ignore that comment.

The PR looks good to me, maybe the committers can comment on it?

imatiach-msft commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -743,6 +743,84 @@ class GeneralizedLinearRegressionSuite
+    var idx = 0
+    for (fitIntercept <- Seq(false, true)) {
+      for (family <- Seq("gaussian", "poisson", "gamma", "tweedie")) {
+        var trainer = new GeneralizedLinearRegression().setFamily(family)
```
just a suggestion: maybe refactor the code below in a method and do fitIntercept.map(fi => family.map(fm => (fi, fm)).zip(expected).foreach(params => callMyMethod(params._1, params._2, params._3))
then you would get rid of the for loops and not have one long test case and you could remove the idx += 1 below

actuaryzhang committed:
cast offset and add in instrumentation

actuaryzhang commented:
@zhengruifeng Thanks for the suggestions. Added casting and instrumentation. 
@imatiach-msft Thanks for the clarification! It is probably worth  another PR to clean up all tests in GLM.  
Let me know if there is any additional comments!

SparkQA commented:
**[Test build #72095 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72095/testReport)** for PR 16699 at commit [`52bc32b`](https://github.com/apache/spark/commit/52bc32b2d86b2cd5ce092f86ee61f8fe9aebec5d).

SparkQA commented:
**[Test build #72095 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72095/testReport)** for PR 16699 at commit [`52bc32b`](https://github.com/apache/spark/commit/52bc32b2d86b2cd5ce092f86ee61f8fe9aebec5d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72095/
Test PASSed.

imatiach-msft commented:
LGTM, thanks!

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -336,14 +361,19 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
-        case Row(label: Double, weight: Double, features: Vector) =>
-          Instance(label, weight, features)
-      }
+    val off = if (!isDefined(offsetCol) || $(offsetCol).isEmpty) {
```
maybe rename off to offset?  also it looks like the spacing below needs to be fixed (extra indentation).
+    val off = if (!isDefined(offsetCol) || $(offsetCol).isEmpty) { 
+      lit(0.0) 
+    } else { 
+      col($(offsetCol)).cast(DoubleType) 
+    }

actuaryzhang replied:
Could you point out which lines have the spacing issue? Thanks.

actuaryzhang committed:
update var name

SparkQA commented:
**[Test build #72162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72162/testReport)** for PR 16699 at commit [`59e10f7`](https://github.com/apache/spark/commit/59e10f77f2676a3755536a46d171ee723d226ee5).

actuaryzhang commented:
@imatiach-msft Thanks much for your review. Renamed `off`.

actuaryzhang commented:
@yanboliang @zhengruifeng @srowen 
Could you guys take a look and let me know if there is any changes needed? Thanks much!

SparkQA commented:
**[Test build #72162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72162/testReport)** for PR 16699 at commit [`59e10f7`](https://github.com/apache/spark/commit/59e10f77f2676a3755536a46d171ee723d226ee5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72162/
Test PASSed.

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -336,14 +361,19 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
+      lit(0.0)
+    } else {
+      col($(offsetCol)).cast(DoubleType)
+    }
```
minor style comment - I think these 4 lines here need to be indented with 2 more spaces:
  lit(0.0)  
} else {  
  col($(offsetCol)).cast(DoubleType)  
}

actuaryzhang replied:
I'm following the style in existing code, for example: 
```
  lazy val rank: Long = if (model.getFitIntercept) {
    model.coefficients.size + 1
  } else {
    model.coefficients.size
  }
```

sethah commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -743,6 +743,84 @@ class GeneralizedLinearRegressionSuite
+        if (family == "tweedie") trainer = trainer.setVariancePower(1.5)
+        val model = trainer.fit(dataset)
+        val actual = Vectors.dense(model.intercept, model.coefficients(0), model.coefficients(1))
+        assert(actual ~= expected(idx) absTol 1e-4, s"Model mismatch: GLM with family = $family," +
```
We need to be checking more than just the coefficients. For example, the computation of the null deviance does not match R, since the null model computation does not consider the offsets.

Actually, I think we ought to just incorporate offsets into all of the other tests, which will make sure offsets are exhaustively tested. This has been done before e.g. https://github.com/apache/spark/pull/15488, and it _is_ a real pain, but it's probably the best way. I'd be open to other arguments though.

actuaryzhang replied:
Thanks for pointing out this. Yes, I also encountered the issue with intercept only + offset model. Will look into this and fix.

actuaryzhang committed:
add test for intercept only

actuaryzhang committed:
update test

actuaryzhang committed:
pull and merge

actuaryzhang committed:
implement null dev for offset model

actuaryzhang committed:
fix null deviance calculation and add tests

SparkQA commented:
**[Test build #72672 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72672/testReport)** for PR 16699 at commit [`fc64d32`](https://github.com/apache/spark/commit/fc64d32e72e1e406d6f7ab27563ea622aa7b1397).

actuaryzhang commented:
@sethah @imatiach-msft 

Please review the new commit. Main changes:
- Fix issue in null deviance calculation in the presence of offset. Except for special cases (Gaussian with identity link), we need to re-estimate a intercept only model. The discussions in #16740 are indeed very helpful here. 
- Updated the test for weighted GLM to include offset as well in order to test the addition of offset thoroughly. 

Please let me know if there is any issue. Thanks!

SparkQA commented:
**[Test build #72672 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72672/testReport)** for PR 16699 at commit [`fc64d32`](https://github.com/apache/spark/commit/fc64d32e72e1e406d6f7ab27563ea622aa7b1397).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72672/
Test PASSed.

sethah commented:
whew, this was a lot of work :)

actuaryzhang commented:
@sethah Yes, that is lots of work. However, the only critical change (since the last commit) is on the calculation of the null deviance. The other changes are mainly because of updating existing tests as you suggested (which makes sense and indeed helps to capture a few issues). This requires adding new offsets and updating all inference results... I think it makes sense to just focus on the null deviance part.

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -1218,16 +1266,35 @@ class GeneralizedLinearRegressionSummary private[regression] (
+      case Row(y: Double, offset: Double, weight: Double) =>
+        family.deviance(y, link.unlink(intercept + offset), weight)
     }.sum()
   }
```
@sethah This part is the most critical change since last time.

sethah commented:
@actuaryzhang This is looking pretty good overall. Regarding the prediction logic, R glm does not allow you to predict with offsets, correct? I notice that statsmodels in Python _does_ allow it. For, e.g. Poisson regression, predicting with offsets is a _slight_ convenience, I guess. The point of training with offsets in this case is to effectively model _rates_ instead of _counts_, and so when you predict you'd be interested in the rates and predicting without offsets, that's exactly what you get. If you want to apply those rates to a specific exposure time, then you just take your predictions and multiply them with `exposure * y_hat` where y_hat is the predicted rate. I think I have that correct. My question then, is do we need to support prediction with offset? It makes things clunkier in the code, so there is some reason not to. Since you presumably have more experience with using GLMs, I'd like to hear your thoughts.

actuaryzhang commented:
@sethah Thanks much for your review. 

Regarding prediction, both R and my implementation here allow prediction with offsets. If the users want to get the predicted rates (instead of counts), then they can specify the offset to be `lit(0.0)` in the new data set when making predictions. 

The following shows that R prediction includes offsets:
```
A <- matrix(c(0, 1, 2, 3, 5, 7, 11, 13), 4, 2)
b <- c(1, 0, 0.2, 2)
off <- c(2, 3, 5, 4)
df <- as.data.frame(cbind(A, b))
model <- glm(formula = "b ~ .", family = "poisson", data = df, 
             offset = off)
p1 <- predict(model, df, type = "response")
p2 <- as.numeric(exp(cbind(1, A) %*% coef(model) + off))
sum(p1 - p2)
```

This part of the code in the GeneralizedLinearRegression file shows that we do need offset when making the prediction. 
```
  protected def predict(features: Vector, offset: Double): Double = {
    val eta = predictLink(features, offset)
    familyAndLink.fitted(eta)
  }
```

And this is the test for the prediction:
``` 
val familyLink = FamilyAndLink(trainer)
        model.transform(dataset).select("features", "offset", "prediction", "linkPrediction")
          .collect().foreach {
          case Row(features: DenseVector, offset: Double, prediction1: Double,
          linkPrediction1: Double) =>
            val eta = BLAS.dot(features, model.coefficients) + model.intercept + offset
            val prediction2 = familyLink.fitted(eta)
            val linkPrediction2 = eta
            assert(prediction1 ~= prediction2 relTol 1E-5, "Prediction mismatch: GLM with " +
              s"family = $family, and fitIntercept = $fitIntercept.")
            assert(linkPrediction1 ~= linkPrediction2 relTol 1E-5, "Link Prediction mismatch: " +
              s"GLM with family = $family, and fitIntercept = $fitIntercept.")
        }
```

Does this make sense?

sethah commented:
I'm finding R's behavior for prediction with offsets to be a bit strange. Yes, R does use the original offsets supplied during training to do prediction, but what if I want to make predictions on a completely new dataset, with completely different offsets? R still tries to use the original ones, which doesn't make sense to me. Then, maybe predicting on "new" data when you've trained using offsets doesn't make sense? Regardless, here we are allowing users to pass in a new dataframe to make predictions on, where they can supply an offset column for the new "test" data - offsets that are not the same as the training offsets. This is a feature not provided by R, it seems.

actuaryzhang commented:
@sethah The predict method can work with new data in R. See below. Shall we focus on the current implementation, instead of discussing the details of the R behavior? :) 
Let me know if there is any suggestion on the last commit. Thanks. 

```
df <- data.frame(y = rnorm(10), x = rnorm(10), off = rnorm(10))
model <- glm(y ~ x, data = df, offset = off)
df2 <- data.frame(y = rnorm(20), x = rnorm(20), off = rnorm(20))
p1 <- predict(model, df2, type = "response")
p2 <- as.numeric(model.matrix(y~x, df2) %*% coef(model) + df2$off)
sum(p1 - p2)
```

sethah commented on mllib/src/main/scala/org/apache/spark/ml/feature/Instance.scala:
```diff
@@ -27,3 +27,25 @@ import org.apache.spark.ml.linalg.Vector
+ * @param offset The offset used for this data point.
+ * @param features The vector of features for this data point.
+ */
+private[ml] case class OffsetInstance(label: Double, weight: Double, offset: Double,
```
I don't have a strong preference, but maybe calling it `GLMInstance` is clearer?

sethah replied:
Also, we should be using Spark style indentation:

````scala
private[ml] case class OffsetInstance(
    label: Double,
    weight: Double,
    offset: Double,
    features: Vector) {
````

actuaryzhang replied:
Updated the style. 
I think I'll stick with OffsetInstance. There may be other models in the future that use offset.

sethah commented:
Ah, thank you very much for that clarification, I don't have much experience using R. I tried this out earlier, and it seems you have to use the same offset column name as you show above. I apologize for not making my intentions clear earlier - I wanted to discuss the R behavior to make sure we were doing the right thing here. 

We actually discuss R's behavior a lot and even use it as a baseline for our unit tests, as you know. I think being able to compare new Spark PRs to other packages enables reviewers who may not be experts in a particular area to still contribute. More importantly, since it takes quite a lot of time and effort to do code review I like to make sure we have decided on the high-level approach before spending time on implementation details, which I find saves time in the long run. JMO, hope it makes sense :)

sethah reviewed (commented):
Made a pass, thanks for working on this!

imatiach-msft commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -798,77 +798,160 @@ class GeneralizedLinearRegressionSuite
+    var idx = 0
+
+    for (fitIntercept <- Seq(false, true)) {
+      for (family <- Seq("gaussian", "binomial", "poisson", "gamma", "tweedie")) {
```
+1 this is a great idea

actuaryzhang replied:
I did implement this, but it seems that the order of the values in `GeneralizedLinearRegression.supportedFamilyNames` changes from test to test... I'm not sure why this happened but since it's a minor issues, I just reverted it back.

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -168,6 +179,7 @@ private[regression] trait GeneralizedLinearRegressionBase extends PredictorParam
     }
 
     val newSchema = super.validateAndTransformSchema(schema, fitting, featuresDataType)
+    if (isSetOffsetCol(this)) SchemaUtils.checkNumericType(schema, $(offsetCol))
```
+1 good catch

actuaryzhang replied:
@sethah This is a great point. The new commit does allow offset to be missing when making predictions. I now check validity of offset only when it's set and available in the prediction set. Otherwise, set offset to be zero. Thanks for catching this and contributing the test.

actuaryzhang committed:
allow missing offset in prediction

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

actuaryzhang committed:
clean up

SparkQA commented:
**[Test build #72857 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72857/testReport)** for PR 16699 at commit [`90d68a6`](https://github.com/apache/spark/commit/90d68a67815aceae63eaad7345477a082bb2febd).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #72858 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72858/testReport)** for PR 16699 at commit [`e95c25b`](https://github.com/apache/spark/commit/e95c25b73682669b65f194141ae08c56deb4d90c).

actuaryzhang commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -798,77 +798,160 @@ class GeneralizedLinearRegressionSuite
+
+    for (fitIntercept <- Seq(false, true)) {
+      for (family <- Seq("gaussian", "binomial", "poisson", "gamma", "tweedie")) {
+        var trainer = new GeneralizedLinearRegression().setFamily(family)
```
Changed.

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -406,6 +435,14 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
+
+  /** Checks whether offset column is set and nonempty */
+  private[regression] def isSetOffsetCol(params: GeneralizedLinearRegressionBase): Boolean =
+    params.isSet(params.offsetCol) && !params.getOffsetCol.isEmpty
```
I adopted `params.getOffsetCol.nonEmpty`. 
As for `isDefined` VS `isSet`, in my other PR #16344, @yanboliang suggested `isSet` is more accurate and we should be using `isSet` and may need to change the existing `isDefined` at some point. For this reason, I'm using `isSet` (which I believe checks whether the param is explicitly set by the user).

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -944,15 +981,27 @@ class GeneralizedLinearRegressionModel private[ml] (
 
   override protected def predict(features: Vector): Double = {
-    val eta = predictLink(features)
+    if (!isSetOffsetCol(this)) {
```
Done.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72857/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72858/
Test FAILed.

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -1139,54 +1189,52 @@ class GeneralizedLinearRegressionSummary private[regression] (
+  private def prediction: Column = col(predictionCol)
+
+  private def weight: Column = {
+    if (!isSetWeightCol(model)) lit(1.0) else col(model.getWeightCol)
```
The computation is the same, but I just changed the name. The original name was kind of confusing: the name `xxxCol` is used to refer the string name in almost all other places, but here it was used as `Column`.  Since I need to add the offset column, I just updated the name  so that they are consistent. Also, the method `isSetWeightCol` is a simple wrapper of the old logic that checks both setting and nonempty of weightCol.  It is also used in a few other places, hence the wrapper.

actuaryzhang commented:
@sethah Thanks much for your review. I've made a new commit that addressed all your comments. Please see my inline comments.  Let me know if there is any other suggestions. Thanks.

actuaryzhang committed:
Merge branch 'master' of https://github.com/apache/spark into offset

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #72863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72863/testReport)** for PR 16699 at commit [`4b336be`](https://github.com/apache/spark/commit/4b336be160fc75c4908ee7ac4a17b3eb6d04541c).

SparkQA commented:
**[Test build #72863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72863/testReport)** for PR 16699 at commit [`4b336be`](https://github.com/apache/spark/commit/4b336be160fc75c4908ee7ac4a17b3eb6d04541c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72863/
Test PASSed.

actuaryzhang commented:
@sethah Is there anything else you would recommend for this PR? Thanks.

actuaryzhang commented:
@yanboliang @sethah 
Any suggestion on moving this PR forward? Appreciate your comments and reviews.

yanboliang commented:
@actuaryzhang Sorry for late response. I will review this PR after finishing 2.2 release, thanks for your patience.

actuaryzhang commented:
@yanboliang Thanks for the update. 
This is a big change, although lots of then are on the test side. 
Let me know if there is anything I can do to help make the review easier.

yanboliang reviewed (commented):
@actuaryzhang Looks good overall, I left some minor comments. Thanks!

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -339,15 +364,16 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
+        }
       val optimizer = new WeightedLeastSquares($(fitIntercept), $(regParam), elasticNetParam = 0.0,
         standardizeFeatures = true, standardizeLabel = true)
       val wlsModel = optimizer.fit(instances)
```
I would suggest we leave `WeightedLeastSquares` as is, since it is a general purpose optimization tool and offset is more specific to GLM. I have not seen a weighted least squares implementation that supports offset. 

We discussed something relevant above [here](https://github.com/apache/spark/pull/16699/files/d44974cfe50092bb639a31aa7aa9b16eb1d21fae#diff-6759d92c079f0957bfa814e339e10e7eR301
). I originally defined `val instances: RDD[OffsetInstance]` outside the `ifelse` and then convert it to `RDD[Instance]` for the Gaussian identity link case. But this will incur one extra map. There was some concern that this could be expensive. However, if this extra conversion is not a big deal, I can revert it back to that which is basically the implementation of the `OffsetInstance` interface for `WeightedLeastSquares`.

actuaryzhang commented on mllib/src/test/scala/org/apache/spark/ml/optim/IterativelyReweightedLeastSquaresSuite.scala:
```diff
@@ -169,29 +169,29 @@ class IterativelyReweightedLeastSquaresSuite extends SparkFunSuite with MLlibTes
+    val eta = model.predict(instance.features) + instance.offset
     val mu = 1.0 / (1.0 + math.exp(-1.0 * eta))
-    val z = eta + (instance.label - mu) / (mu * (1.0 - mu))
+    val z = eta - instance.offset + (instance.label - mu) / (mu * (1.0 - mu))
```
Indeed this is the correct implementation: in the IRWLS, we only include offset when computing `mu` and use `Xb` (without offset) when updating the working label. To see this clearly, one would have to derive the IRWLS. But for a quick reference, below is R's implementation:

```
eta <- drop(x %*% start)
mu <- linkinv(eta <- eta + offset)
z <- (eta - offset)[good] + (y - mu)[good]/mu.eta.val[good]
w <- sqrt((weights[good] * mu.eta.val[good]^2)/variance(mu)[good])
fit <- .Call(C_Cdqrls, x[good, , drop = FALSE] * 
              w, z * w, min(1e-07, control$epsilon/1000), check = FALSE)
```

actuaryzhang commented on mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:
```diff
@@ -798,77 +798,184 @@ class GeneralizedLinearRegressionSuite
-       1.92 -0.96 -0.64  0.48
-
-       residuals(model, type="response")
+      0.96 -0.48 -0.32  0.24
```
They don't have the same indentation with the old code but within the new block of code, every line is still aligned. Won't change.

actuaryzhang committed:
address comments

actuaryzhang commented:
@yanboliang Thanks much for the review. The new commit includes everything you suggested except implementing `WeightLeastSquares` interface for `OffsetInstance`. Please see my incline comments above.

SparkQA commented:
**[Test build #78727 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78727/testReport)** for PR 16699 at commit [`1e47a11`](https://github.com/apache/spark/commit/1e47a11f6a3e1292426ea25e7728b805c6650e1b).

SparkQA commented:
**[Test build #78727 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78727/testReport)** for PR 16699 at commit [`1e47a11`](https://github.com/apache/spark/commit/1e47a11f6a3e1292426ea25e7728b805c6650e1b).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78727/
Test FAILed.

actuaryzhang commented:
jenkins, retest this please

SparkQA commented:
**[Test build #78729 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78729/testReport)** for PR 16699 at commit [`1e47a11`](https://github.com/apache/spark/commit/1e47a11f6a3e1292426ea25e7728b805c6650e1b).

SparkQA commented:
**[Test build #78729 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78729/testReport)** for PR 16699 at commit [`1e47a11`](https://github.com/apache/spark/commit/1e47a11f6a3e1292426ea25e7728b805c6650e1b).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78729/
Test FAILed.

actuaryzhang commented:
Not sure what this error msg means, but it seems unrelated to this PR.

felixcheung commented:
this is a known issue in test runs currently - it's mentioned in dev@spark.apache.org, just so you know.

actuaryzhang commented:
Got it. I should pay more attention to that mailing list from now on :)

yanboliang commented:
Jenkins, test this please.

SparkQA commented:
**[Test build #78812 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78812/testReport)** for PR 16699 at commit [`1e47a11`](https://github.com/apache/spark/commit/1e47a11f6a3e1292426ea25e7728b805c6650e1b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78812/
Test PASSed.

yanboliang reviewed (commented):
LGTM after resolving some trivial issues. Thanks for contributing this great feature.

actuaryzhang committed:
address comments

actuaryzhang commented:
Made a new commit that fixes the issues you pointed out.

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala:
```diff
@@ -961,14 +1008,16 @@ class GeneralizedLinearRegressionModel private[ml] (
   }
 
   override protected def transformImpl(dataset: Dataset[_]): DataFrame = {
```
Thanks for summarizing the different cases. I think this is worth a deeper discussion as follow-up work. Let me work on this in another PR.

SparkQA commented:
**[Test build #78920 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78920/testReport)** for PR 16699 at commit [`db0ac93`](https://github.com/apache/spark/commit/db0ac937b9df678f7f58b26bab1afa43d77eb5a1).

SparkQA commented:
**[Test build #78920 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78920/testReport)** for PR 16699 at commit [`db0ac93`](https://github.com/apache/spark/commit/db0ac937b9df678f7f58b26bab1afa43d77eb5a1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78920/
Test PASSed.

yanboliang commented:
Merged into master. Thanks for contribution and all reviews! This great feature will benefit lots of users.
@actuaryzhang Could you send follow-up PRs to address the two inline comments? Thanks.

hvanhovell commented:
@yanboliang @actuaryzhang this PR breaks the scala-2.10 build:
```
[error] /home/jenkins/workspace/spark-master-compile-maven-scala-2.10/mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:1117: value absTol is not a member of Int
[error]     assert(summary.dispersion ~== dispersionR absTol 1E-3)
[error]                                               ^
[error] /home/jenkins/workspace/spark-master-compile-maven-scala-2.10/mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala:1232: value absTol is not a member of Int
```
See: https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-master-compile-maven-scala-2.10/4721/

yanboliang commented:
@hvanhovell I will send a quick fix soon, thanks for your kindly remind.

yanboliang commented:
#18489 fixed the build failure. Thanks.
